processing/src/main/java/io/druid/query/QueryMetricsFactory.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import io.druid.query.groupby.GroupByQueryMetrics;
+import io.druid.query.timeseries.TimeseriesQueryMetrics;
+import io.druid.query.topn.TopNQueryMetrics;
+
+/**
+ * This factory is used for DI of custom {@link QueryMetrics} implementations.
+ */
+public interface QueryMetricsFactory
+{
+  /**
+   * Creates a generic {@link QueryMetrics} for query, which doesn't require a custom QueryMetrics subclass (i. e. not
+   * a topN, groupBy or timeseries query). This method must call {@link QueryMetrics#query(Query)} with the given query
+   * on the created QueryMetrics object before returning.
+   */
+  QueryMetrics<Query<?>> makeMetrics(Query<?> query);
```

Comment:
If someone wants to use the same dimensions for scan query, but emit them differently, i. e. more or less verbose than default, he should implement his own `QueryMetricFactory` with `makeMetrics()` similar to:
```
QueryMetrics makeMetrics(Query q)
{
    QueryMetrics metrics = q instanceof ScanQuery ? new CustomScanQueryMetrics() : new DefaultQueryMetrics();
    metrics.query(q);
    return metrics;
}
```

If `ScanQuery`  wants to emit unique dimensions, I think it needs to make an own subclass of `QueryMetrics`, the same way topN, groupBy and timeseries already go. Since it's an extension, I don't think it should make a new method in `QueryMetricsFactory`.

On second thought it seems unnecessarily complex and creating undesired dependencies between components to require extensions to inject though the general `QueryMetricsFactory`. We may require extensions to provide their own means to injecting specialized implementations of their `QueryMetrics` subclasses. I. e. if `ScanQuery` needs `ScanQueryMetrics`, it should also make an injectable `ScanMetricsQueryFactory` and allow users to overwrite it. What do you think?

<hr>

Regarding `userDimensions()`, it was actually an artifact of the old string-based approach, removed it.